### ISSUE TITLE
issues when converting from haskell numbers to typed tensors

### ISSUE THREAD

**Issue by tscholak** [hasktorch/hasktorch#429]
This shouldn't be:
```
> ((fromIntegral 2) - 1.0) :: Tensor '( 'CPU, 0) 'Int64 '[3,3]
Tensor Float []  1.0000
```
neither runtime dtype nor shape match the respective types.

**Comment by tscholak** [comment]
to be fair, unless we do some automated broadcasting, I don't think the conversion from `Num` to anything but `shape ~ '[]` shouldn't be allowed.

**Comment by RichardWarfield** [comment]
This bit me today.  I tend to think broadcasting is the right thing to do here - we essentially would have `fromInteger` and `fromRational` as synonyms for `full`.  Although I agree that it would be nice to restrict these functions to `'[]` tensors, I don't think there is any way to do this without breaking the other arithmetic operators in the `Num` class.

**Comment by tscholak** [comment]
It’s an odd response, but thanks for running into this and reminding me of this issue. I’m going to fix this for the gradually typed API. Because `fromInteger` doesn’t really make sense for tensors with arbitrary dimensions, I was about to give up on defining a `Num` instance. I’m going to use `full` like you suggested!

**Comment by RichardWarfield** [comment]
I thought of a (maybe) little caveat to to this approach.  Consider code like the following:

```haskell
f :: Tensor '(CUDA, 0) Float '[1000,10000]
f a = a +1
```

hasktorch is going to allocate a huge CUDA tensor just to add a scalar to a tensor.  Actually it's much worse than that - GHC will very likely float out the constant tensor to the top level, meaning it never gets deallocated.  

I still think the `Num` instance for `Tensor` is too nice to give up, though...

**Comment by tscholak** [comment]
we could use `expand` instead, but that may cause other problems: https://pytorch.org/docs/stable/tensors.html?highlight=repeat#torch.Tensor.expand because the tensor of ones won't be contiguous.

**Comment by tscholak** [comment]
@RichardWarfield 
I've just hacked together this funky instance:
https://github.com/hasktorch/hasktorch/blob/gradually-typed-hasktorch/hasktorch/src/Torch/GraduallyTyped/Tensor/Type.hs#L129-L170
There are a lot of possibilities for runtime errors.

**Comment by tscholak** [comment]
Of course, one could just create the integer as a zero-dimensional tensor with default tensor opts and then hope that broadcasting and runtime unification just works out. Maybe that's what's happening in PyTorch.

**Comment by RichardWarfield** [comment]
> we could use `expand` instead, but that may cause other problems: https://pytorch.org/docs/stable/tensors.html?highlight=repeat#torch.Tensor.expand because the tensor of ones won't be contiguous.

Of course!  `expand` is a great idea.  I don't think the tenor being non-contiguous is likely to cause problems in practice.  As a user I would hardly expect a numeric literal to allocate a big continuous tensor.

**Comment by tscholak** [comment]
> I don't think the tensor being non-contiguous is likely to cause problems in practice.

**Comment by RichardWarfield** [comment]
Sure, but how likely are those ops to be used on tensors created as numeric literals?  Granted, it would be a pretty annoying problem to debug.  But I can't think of a context in which it is actually likely to arise.
